File: ios/Classes/Methods/UserMethods.swift
import Flutter
import Foundation
import TwilioConversationsClient

public class UserMethods {
    public static func unsubscribe(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any?] else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing parameters", details: nil))
        }

        guard let identity = arguments["identity"] as? String else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing 'identity' parameter", details: nil))
        }

        let flutterResult = result
        SwiftTwilioConversationsPlugin.chatListener?.chatClient?.subscribedUser(
            withIdentity: identity, completion: { (result, user) in
                if result.isSuccessful {
                    user?.unsubscribe()
                    flutterResult(nil)
                } else {
                    flutterResult(FlutterError(code: "ERROR",
                        message: "No subscribed user found with the identity '\(identity)'", details: nil))
                }
        })
    }
}

File: flutter_twilio_conversations/android/src/main/kotlin/twilio/flutter/twilio_conversations/methods/UserMethods.kt
package twilio.flutter.twilio_conversations.methods

import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import twilio.flutter.twilio_conversations.TwilioConversationsPlugin

object UserMethods {
    fun unsubscribe(call: MethodCall, result: MethodChannel.Result) {
        val identity = call.argument<String>("identity")
                ?: return result.error("ERROR", "Missing 'identity'", null)

        try {
            val subscribedUser = TwilioConversationsPlugin.chatClient?.subscribedUsers?.find { it.identity == identity }

            if (subscribedUser != null) {
                subscribedUser.unsubscribe()
            } else {
                return result.error("ERROR", "No subscribed user found with the identity '$identity'", null)
            }
        } catch (err: IllegalArgumentException) {
            return result.error("IllegalArgumentException", err.message, null)
        }
    }
}
